Message bus plugin initial implementation

Jurisak opened the pull request:
This commit also contains ansible module in Library directory

Commit:
Message bus plugin initial implementation
This commit also contains ansible module in Library directory

gsr-shanks commented on plugins/library/messagehub.py:
```diff
@@ -0,0 +1,81 @@
+import logging.config
+from ansible.module_utils.basic import *
+
+
```
Could you please follow: http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html

Commit:
Added documentation into ansible module

Commit:
Added documentation into ansible module

gsr-shanks commented on plugins/library/messagehub.py:
```diff
@@ -0,0 +1,161 @@
+    description:
+      - Message bus host.
+    required: false
+    default: ci-bus.lab.eng.rdu2.redhat.com
```
We should not set a default value and change 'required:' to 'True'. Since the user should be able to provide whichever message bus host to listen on.

Jurisak replied:
As far as I see required value is set to false.

gsr-shanks replied:
Right, and it should be required.

Jurisak replied:
So you mean to erase default value and change required to True?

gsr-shanks replied:
Yes. It is internal hostname anyway.

gsr-shanks commented on plugins/library/messagehub.py:
```diff
@@ -0,0 +1,161 @@
+
+EXAMPLES = '''
+- name: Get single message from message bus
+  messagehub:
```
Add 'host:' here ?

Jurisak replied:
Right ... :+1:

Jurisak replied:
Please do a full review and I will push all changes in a single commit.

gsr-shanks commented on plugins/morph_messagehub.py:
```diff
@@ -0,0 +1,112 @@
+    """Parse command line arguments."""
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        description='Subscribe to the Red Hat CI message bus.'
```
' Red Hat CI' is not required in the description here.

gsr-shanks commented on plugins/morph_messagehub.py:
```diff
@@ -0,0 +1,112 @@
+    if listener.error_message:
+        exit("Got error message through message bus {0}".format(listener.error_message))
+    with open("metamorph.json", "w") as metamorph:
+        json.dump(dict(messages=listener.metamorph_data), metamorph)
```
Would be nice to have this in pretty JSON.

gsr-shanks commented on plugins/morph_messagehub.py:
```diff
@@ -0,0 +1,112 @@
+            self.count += 1
+
+
+def main():
```
Why no logging here ?

gsr-shanks commented on plugins/library/messagehub.py:
```diff
@@ -0,0 +1,161 @@
+    conn.disconnect()
+
+    if not listener.error_message:
+        module.exit_json(changed=True, meta=dict(messages=listener.metamorph_data))
```
Can we have listener.metamorph_data dumped to metamorph.json along with module.exit_json ?

Commit:
Added logging and json pretty print

gsr-shanks commented on plugins/morph_messagehub.py:
```diff
@@ -0,0 +1,137 @@
+            config = json.load(f)
+        logging.config.dictConfig(config)
+    else:
+        logging.basicConfig(level=default_level)
```
Can we please have logging in some shared lib ? Since we will need this for all the modules in metamorph.

Jurisak replied:
Any advice's where to put this logging file?

gsr-shanks replied:
Based on google doc, let's have this in 'metamorph/lib'.

gsr-shanks commented:
There are 2 use cases with messagehub.
1. It should be able to listen to messages on the message bus.
2. It should extract the message if the listening part is taken care by any Jenkins plugin.

This PR covers use case 1. Assuming 'CI_MESSAGE' is available in the env variable we should dump that into metamorph.json.

Jurisak commented:
Ok so you are suggesting that If there is a existing ENV variable which contains CI_MESSAGE then Metamorph will only take the message in dump it into metamorph.json?

gsr-shanks commented:
Yes.

gsr-shanks commented on logging.json:
```diff
@@ -0,0 +1,51 @@
+        "level": "INFO",
+        "handlers": ["console", "debug_file_handler", "error_file_handler"]
+    }
+}
```
This file should be in metamorph/etc/

Commit:
Added new env-option into message bus plugin logic

This commit also include creation of two directories. Lib: for logging
and etc: for logging.json

Commit:
Added new env-option into message bus plugin logic

This commit also include creation of two directories. Lib: for logging
and etc: for logging.json

gsr-shanks commented:
Please remove *.pyc files. Even better if you can include .gitingnore.

Jurisak commented:
Sure ;). Sorry ..

Commit:
Deleted .pyc files

Commit:
Deleted .pyc files

gsr-shanks commented:
[root@00bee40107d6 metamorph]# python setup.py install
  File "setup.py", line 9
SyntaxError: Non-ASCII character '\xc5' in file setup.py on line 9, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details
[root@00bee40107d6 metamorph]#

Commit:
Added fix in setup.py file

Commit:
Merge branch 'messagebus_reader' of github.com:Jurisak/metamorph into messagebus_reader

Jurisak commented:
@gsr-shanks Is it better now?

gsr-shanks commented:
Yes, it is better now. Thanks.

gsr-shanks commented:
Please add documentation and an option to provide output file name and default to metamorph.json if none.

Commit:
Added support of different json output names.

This option is supported by "option" command

Commit:
Added support of different json output names.

This option is supported by "option" command

gsr-shanks commented on plugins/library/messagehub.py:
```diff
@@ -75,6 +81,12 @@
+- name: Get single message from environmental variable and store it into hello.json
+  messagehub:
+    env-variable: "..."
+    output "hello.json"
```
You mean: 

`output: "hello.json"` ?

Jurisak replied:
yes

gsr-shanks commented on lib/logging_conf.py:
```diff
@@ -0,0 +1,29 @@
+        logging.basicConfig(level=default_level)
+
+
+def storing_pretty_json(input_data, output="metamorph.json"):
```
output shouldn't be a str here. right ?

gsr-shanks commented on plugins/library/messagehub.py:
```diff
@@ -0,0 +1,202 @@
+        ci_message = os.environ.get(module.params['env-variable'], "UNKNOWN")
+        if ci_message == "UNKNOWN":
+            logging.error("Environmental variable didn't found")
+            error_message = "Environmental variable didn't found"
```
s/Environmental variable didn't found/Environment variable not found/g

gsr-shanks replied:
s/Environmental variable didn't found/Environment variable not found/g

gsr-shanks commented on lib/logging_conf.py:
```diff
@@ -0,0 +1,29 @@
+        logging.basicConfig(level=default_level)
+
+
+def storing_pretty_json(input_data, output="metamorph.json"):
```
output shouldn't be a str here. right ?

Jurisak replied:
Why not?

gsr-shanks replied:
I mean it is hardcoded to 'metamorph.json' file name.

Jurisak replied:
This is only default value. I'm passing different one. I thought that is better to have a good default instead of nothing...

gsr-shanks replied:
But isn't the default handled already if the file name arg is not provided ?

Jurisak replied:
Sure, but we would probably use this function in following metamorph plugins and the default would be "metamorph.json" as well ...

gsr-shanks replied:
Ack!

gsr-shanks commented on plugins/library/messagehub.py:
```diff
@@ -0,0 +1,202 @@
+        )
+    logging.info("Connection to message bus established.")
+    while (not listener.error_message) and len(listener.metamorph_data) < module.params['count']:
+        logging.info("Waiting 1s for CI message to arrive")
```
Can we have this as logging.debug ? logging.info for every 1s is too much of an output.

gsr-shanks commented on plugins/library/messagehub.py:
```diff
@@ -0,0 +1,202 @@
+            id='1',
+            ack='auto'
+        )
+    logging.info("Connection to message bus established.")
```
Probably add:
`logging.info("Waiting for CI message to arrive ...")`

gsr-shanks commented on plugins/morph_messagehub.py:
```diff
@@ -0,0 +1,162 @@
+            id='1',
+            ack='auto'
+        )
+    logging.info("Connection to message bus established.")
```
same as above.

gsr-shanks commented on plugins/morph_messagehub.py:
```diff
@@ -0,0 +1,162 @@
+        )
+    logging.info("Connection to message bus established.")
+    while (not listener.error_message) and len(listener.metamorph_data) < args.count:
+        logging.info("Waiting 1s for CI message to arrive")
```
same as above.

gsr-shanks commented on plugins/morph_messagehub.py:
```diff
@@ -0,0 +1,162 @@
+def env_run(args):
+    env_data = os.environ.get(args.env_variable, "UNKNOWN")
+    if env_data == "UNKNOWN":
+        logging.error("Environmental variable didn't found")
```
Same as above

gsr-shanks commented on plugins/library/messagehub.py:
```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/python
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
```
version: 0.1 ?

Commit:
Code refactorization

gsr-shanks commented:
Would you submitting .rst docs in this commit ?

Jurisak commented:
I'll do them today. But it depends on when you want to merge this pull request ...

gsr-shanks commented:
Sure, I will have it merged immediately after documentation is submitted.

Jurisak commented:
What do you want to have it in the documentation?

gsr-shanks commented:
Need documentation on how:
 
- To install metamorph
- To use in Ansible playbook
- To execute CLI

Commit:
First implementation of sphinx documentation

gsr-shanks commented:
Please include Makefile to build html docs

Commit:
Sphinx Makefile
